File: backend/app/services/workflow_service.py
from datetime import datetime

from app.core.exceptions import ValidationError
from app.models.project import Project
from app.models.project_version import ProjectVersion
from app.models.workflow_run import WorkflowRun
from app.models.workflow_run_node import WorkflowRunNode
from app.workers.run_tasks import execute_run
from sqlalchemy.orm import Session

SUPPORTED_NODE_TYPES = {"input_video", "prompt_input", "kie_video_task", "output_video"}


class WorkflowService:
    @staticmethod
    def create_and_run(db: Session, user_id: int, project: Project) -> WorkflowRun:
        version = db.get(ProjectVersion, project.latest_version_id) if project.latest_version_id else None
        if not version:
            raise ValidationError("请先保存画布")

        canvas = version.canvas_json or {}
        nodes = canvas.get("nodes", [])

        run = WorkflowRun(
            user_id=user_id,
            project_id=project.id,
            version_id=version.id,
            status="queued",
            input_json={"canvas_node_count": len(nodes)},
            started_at=datetime.utcnow(),
        )
        db.add(run)
        db.flush()

        for node in nodes:
            node_type = node.get("type")
            if node_type not in SUPPORTED_NODE_TYPES:
                continue
            run_node = WorkflowRunNode(
                run_id=run.id,
                node_id=node.get("id", ""),
                node_type=node_type,
                status="queued",
                input_json=node.get("data", {}),
            )
            db.add(run_node)

        db.commit()
        db.refresh(run)
        execute_run.delay(run.id)
        return run

File: backend/src/main/java/com/aivideo/canvas/service/WorkflowService.java
package com.aivideo.canvas.service;

import com.aivideo.canvas.common.AppException;
import com.aivideo.canvas.entity.Project;
import com.aivideo.canvas.entity.ProjectVersion;
import com.aivideo.canvas.entity.WorkflowRun;
import com.aivideo.canvas.entity.WorkflowRunNode;
import com.aivideo.canvas.repository.ProjectVersionRepository;
import com.aivideo.canvas.repository.WorkflowRunNodeRepository;
import com.aivideo.canvas.repository.WorkflowRunRepository;
import com.aivideo.canvas.worker.RunWorker;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;
import java.util.Set;

@Service
public class WorkflowService {
    private static final Set<String> SUPPORTED = Set.of(
            // 原子化 Data Flow 节点
            "text", "image", "audio", "video_gen",
            // 旧节点类型（兼容）
            "input_video", "prompt_input", "kie_video_task", "output_video"
    );
    private final ProjectVersionRepository versionRepository;
    private final WorkflowRunRepository runRepository;
    private final WorkflowRunNodeRepository nodeRepository;
    private final RunWorker runWorker;
    private final ObjectMapper objectMapper = new ObjectMapper();

    public WorkflowService(ProjectVersionRepository versionRepository, WorkflowRunRepository runRepository, WorkflowRunNodeRepository nodeRepository, RunWorker runWorker) {
        this.versionRepository = versionRepository;
        this.runRepository = runRepository;
        this.nodeRepository = nodeRepository;
        this.runWorker = runWorker;
    }

    public WorkflowRun createAndRun(Long userId, Project project){
        if (project.getLatestVersionId() == null) throw new AppException("VALIDATION_ERROR","请先保存画布");
        ProjectVersion version = versionRepository.findById(project.getLatestVersionId()).orElseThrow(() -> new AppException("NOT_FOUND","版本不存在"));
        WorkflowRun run = new WorkflowRun();
        run.setUserId(userId); run.setProjectId(project.getId()); run.setVersionId(version.getId()); run.setStatus("queued");
        run.setTriggerSource("manual"); run.setInputJson("{}"); run.setOutputJson("{}"); run.setStartedAt(LocalDateTime.now()); run.setErrorMessage("");
        run = runRepository.save(run);

        try {
            Map<String,Object> canvas = objectMapper.readValue(version.getCanvasJson(), new TypeReference<>(){});
            for (Map<String,Object> node : (List<Map<String,Object>>) canvas.getOrDefault("nodes", List.of())) {
                String type = String.valueOf(node.getOrDefault("type", ""));
                if (!SUPPORTED.contains(type)) continue;
                WorkflowRunNode rn = new WorkflowRunNode();
                rn.setRunId(run.getId()); rn.setNodeId(String.valueOf(node.getOrDefault("id", ""))); rn.setNodeType(type); rn.setStatus("queued");
                rn.setInputJson(objectMapper.writeValueAsString(node.getOrDefault("data", Map.of()))); rn.setOutputJson("{}"); rn.setProvider(""); rn.setProviderTaskId(""); rn.setErrorMessage("");
                nodeRepository.save(rn);
            }
        } catch (Exception e){ throw new AppException("VALIDATION_ERROR", e.getMessage()); }

        runWorker.execute(run.getId());
        return run;
    }
}
